modules/renter/proto/contractset.go

```diff
@@ -0,0 +1,159 @@
+package proto
+
+import (
+	"sync"
+
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// A safeContract protects a RenterContract with a mutex.
+type safeContract struct {
+	modules.RenterContract
+	mu sync.Mutex
+}
+
+// A ContractSet provides safe concurrent access to a set of contracts. Its
+// purpose is to serialize modifications to individual contracts, as well as
+// to provide operations on the set as a whole.
+type ContractSet struct {
+	contracts map[types.FileContractID]*safeContract
+	mu        sync.Mutex
+
+	// caches (initialized by initCache)
+	flatIDs       []types.FileContractID
+	flatContracts []modules.RenterContract
+	once          sync.Once
```

Comment:
I think the cache here is over-optimization. Definitely increases the complexity, and since you have to reset the cache frequently anyway, I don't think it really saves all that much. I think the deadlocks are resolved by releasing the maplock before you acquire the safeContract locks, which means you avoid nested lock calls.

My preference I think would be to remove the cache and just iterate through the map. You have to hold the map lock while you do the contract copying, but the deadlocks should be resolved.